e2e/test/FaultInjection.cs

```diff
@@ -62,5 +63,121 @@ public static Client.Message ComposeErrorInjectionProperties(string faultType, s
                 }
             };
         }
+
+        public static async Task ActivateFaultInjection(Client.TransportType transport, string faultType, string reason, int delayInSec, int durationInSec, DeviceClient deviceClient)
+        {
+            s_log.WriteLine($"{nameof(ActivateFaultInjection)}: Requesting fault injection type={faultType} reason={reason}, delay={delayInSec}s, duration={FaultInjection.DefaultDurationInSec}s");
+
+            uint oldTimeout = deviceClient.OperationTimeoutInMilliseconds;
+
+            try
+            {
+                // For MQTT FaultInjection will terminate the connection prior to a PUBACK 
+                // which leads to an infinite loop trying to resend the FaultInjection message.
+                if (transport == Client.TransportType.Mqtt ||
+                    transport == Client.TransportType.Mqtt_Tcp_Only ||
+                    transport == Client.TransportType.Mqtt_WebSocket_Only)
+                {
+                    deviceClient.OperationTimeoutInMilliseconds = 1000;
+                }
+
+                await deviceClient.SendEventAsync(
+                    FaultInjection.ComposeErrorInjectionProperties(
+                        faultType,
+                        reason,
+                        delayInSec,
+                        durationInSec)).ConfigureAwait(false);
+            }
+            catch (TimeoutException ex)
+            {
+                s_log.WriteLine($"{nameof(ActivateFaultInjection)}: {ex}");
+            }
+            finally
+            {
+                deviceClient.OperationTimeoutInMilliseconds = oldTimeout;
+                s_log.WriteLine($"{nameof(ActivateFaultInjection)}: Fault injection requested.");
+            }
+        }
+
+        public static async Task TestErrorInjectionTemplate(
+            string devicePrefix,
+            TestDeviceType type,
+            Client.TransportType transport,
+            string faultType,
+            string reason,
+            int delayInSec,
+            int durationInSec,
+            Func<DeviceClient, TestDevice, Task> initOperation,
+            Func<DeviceClient, TestDevice, Task> testOperation,
```

Comment:
rename "testOperation" to "baselineOperation" for clarity. We aren't testing that basic twin/methods/events scenarios work, we're testing connection recovery.

Reply:
testOperation is used for both baseline and the recovery test, not just baseline.